keel-core/src/test/kotlin/com/netflix/spinnaker/keel/verification/VerificationRunnerTests.kt

```diff
@@ -0,0 +1,234 @@
+  }
+
+  private val subject = VerificationRunner(
+    repository,
+    listOf(evaluator)
+  )
+
+  @Test
+  fun `no-ops for an environment with no verifications`() {
+    val context = VerificationContext(
+      deliveryConfig = DeliveryConfig(
+        application = "fnord",
+        name = "fnord-manifest",
+        serviceAccount = "jamm@illuminati.org",
+        artifacts = setOf(
+          DockerArtifact("fnord")
+        ),
+        environments = setOf(
+          Environment(name = "test")
+        )
+      ),
+      environmentName = "test",
+      version = "fnord-0.190.0-h378.eacb135"
+    )
+
+    subject.runVerificationsFor(context)
+
+    verify(exactly = 0) {
+      evaluator.start(any())
+    }
+  }
+
+  @Test
+  fun `starts the first verification if none have been run yet`() {
+    val context = VerificationContext(
+      deliveryConfig = DeliveryConfig(
+        application = "fnord",
+        name = "fnord-manifest",
+        serviceAccount = "jamm@illuminati.org",
+        artifacts = setOf(
+          DockerArtifact("fnord")
+        ),
+        environments = setOf(
+          Environment(
+            name = "test",
+            verifyWith = setOf(DummyVerification("1"), DummyVerification("2"))
+          )
+        )
+      ),
+      environmentName = "test",
+      version = "fnord-0.190.0-h378.eacb135"
+    )
+
+    every {
+      repository.getState(any(), any())
+    } returns null
+
+    subject.runVerificationsFor(context)
+
+    verify { evaluator.start(DummyVerification("1")) }
+    verify { repository.updateState(any(), DummyVerification("1"), RUNNING) }
+  }
+
+  @Test
+  fun `no-ops if any verification was already running and has yet to complete`() {
+    val context = VerificationContext(
+      deliveryConfig = DeliveryConfig(
+        application = "fnord",
+        name = "fnord-manifest",
+        serviceAccount = "jamm@illuminati.org",
+        artifacts = setOf(
+          DockerArtifact("fnord")
+        ),
+        environments = setOf(
+          Environment(
+            name = "test",
+            verifyWith = setOf(DummyVerification("1"), DummyVerification("2"))
+          )
+        )
+      ),
+      environmentName = "test",
+      version = "fnord-0.190.0-h378.eacb135"
+    )
+
+    every {
+      repository.getState(any(), DummyVerification("1"))
+    } returns RUNNING.toState()
+    every {
+      repository.getState(any(), DummyVerification("2"))
+    } returns null
+
+    every {
+      evaluator.evaluate()
+    } returns RUNNING
+
+    subject.runVerificationsFor(context)
+
+    verify {
+      evaluator.evaluate()
+    }
+    verify(exactly = 0) {
+      evaluator.start(any())
+    }
+  }
+
+  @ParameterizedTest(
+    name = "continues to the next if any verification was already running and has now completed with {0}"
+  )
+  @EnumSource(
+    value = VerificationStatus::class,
+    names = ["PASSED", "FAILED"]
```

Comment:
I assumed we would _not_ run a downstream verification if the previous one failed, i.e. that we would short-circuit in that case. What's the rationale for running it anyway?

Reply:
Avoiding playing whack-a-mole where you fix one thing only to realize the next is broken

Reply:
In that case, shouldn't they run in parallel?

Reply:
Still the subject of debate. Lorin & I thought running sequentially was the simplest thing to start with.